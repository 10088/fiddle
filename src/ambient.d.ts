declare module 'monaco-loader';
declare module 'electron-default-menu';
