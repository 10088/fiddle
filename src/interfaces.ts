export interface StringMap<T> {
  [x: string]: T;
}

export type ElectronVersionState = 'downloading' | 'ready' | 'unknown';

export interface GitHubVersion {
  url: string;
  assets_url: string;
  html_url: string;
  tag_name: string;
  target_commitish: string;
  name: string;
  prerelease: boolean;
  created_at: string;
  published_at: string;
  body: string;
}

export interface ElectronVersion extends GitHubVersion {
  state: ElectronVersionState
}

export interface OutputEntry {
  text: string;
  timestamp: number;
}