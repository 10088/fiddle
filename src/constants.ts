import * as electronModule from 'electron';

const electron = electronModule.remote
  ? electronModule.remote
  : electronModule;

export const USER_DATA_PATH = electron.app.getPath('userData');
export const INDEX_HTML_NAME = 'index.html';
export const MAIN_JS_NAME = 'main.js';
export const RENDERER_JS_NAME = 'renderer.js';
export const PACKAGE_NAME = 'package.json';
export const UNTITLED_NAME = 'untitled-fiddle';

