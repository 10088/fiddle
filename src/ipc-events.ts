export enum IpcEvents {
  MAIN_WINDOW_READY_TO_SHOW = 'MAIN_WINDOW_READY_TO_SHOW',
  OPEN_SETTINGS = 'OPEN_SETTINGS',
  LOAD_GIST_REQUEST = 'LOAD_GIST_REQUEST',
  RUN_FIDDLE = 'RUN_FIDDLE',
  MONACO_EXECUTE_COMMAND = 'MONACO_EXECUTE_COMMAND',
  FS_NEW_FIDDLE = 'FS_NEW_FIDDLE',
  FS_OPEN_FIDDLE = 'FS_OPEN_FIDDLE',
  FS_SAVE_FIDDLE = 'FS_SAVE_FIDDLE',
  FS_SAVE_FIDDLE_AS = 'FS_SAVE_FIDDLE_AS',
  FS_SAVE_FIDDLE_GIST = 'FS_SAVE_FIDDLE_GIST',
  FS_SAVE_FIDDLE_DIALOG = 'FS_SAVE_FIDDLE_DIALOG',
  FS_SAVE_FIDDLE_ERROR = 'FS_SAVE_FIDDLE_ERROR',
}

export const ipcMainEvents = [
  IpcEvents.MAIN_WINDOW_READY_TO_SHOW,
  IpcEvents.FS_SAVE_FIDDLE_DIALOG,
  IpcEvents.FS_SAVE_FIDDLE
];

export const ipcRendererEvents = [
  IpcEvents.OPEN_SETTINGS,
  IpcEvents.LOAD_GIST_REQUEST,
  IpcEvents.RUN_FIDDLE,
  IpcEvents.MONACO_EXECUTE_COMMAND,
  IpcEvents.FS_NEW_FIDDLE,
  IpcEvents.FS_OPEN_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE_AS,
  IpcEvents.FS_SAVE_FIDDLE_GIST,
];
