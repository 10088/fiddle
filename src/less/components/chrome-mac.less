.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
