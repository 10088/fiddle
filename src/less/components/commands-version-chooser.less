button.version-chooser.bp3-button {
  border-radius: 3px;
}
