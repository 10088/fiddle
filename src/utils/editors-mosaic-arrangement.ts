import { MosaicNode } from 'react-mosaic-component';

import { EditorId } from '../interfaces';
import { DEFAULT_MOSAIC_ARRANGEMENT } from '../renderer/state';

/**
 * Create a mosaic arrangement given an array of editor ids.
 *
 * @export
 * @param {Array<EditorId>} input
 * @returns {MosaicNode<EditorId>}
 */
export function createMosaicArrangement(input: Array<EditorId>): MosaicNode<EditorId> {
  if (input.length === 2) {
    return {
      direction: 'column',
      first: input[0],
      second: input[1]
    };
  }

  if (input.length === 3) {
    return DEFAULT_MOSAIC_ARRANGEMENT;
  }

  return input[0];
}

/**
 * Returns an array of visible editors given a mosaic arrangement.
 *
 * @export
 * @param {MosaicNode<EditorId> | null} input
 * @returns {Array<EditorId>}
 */
export function getVisibleEditors(input: MosaicNode<EditorId> | null): Array<EditorId> {
  // Handle the unlikely null case
  if (!input) return [];

  // Handle the case where only one editor is visible
  if (typeof input === 'string') {
    return [ input ];
  }

  // Handle the other cases (2 or 3)
  const result = [];
  for (const node of [ input.first, input.second ]) {
    if (typeof node === 'string') {
      result.push(node);
    } else {
      result.push(...getVisibleEditors(node));
    }
  }

  return result;
}
