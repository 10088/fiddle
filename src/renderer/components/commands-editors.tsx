import { Button, Menu, MenuItem, Popover, Position } from '@blueprintjs/core';
import { observer } from 'mobx-react';
import * as React from 'react';

import { EditorId } from '../../interfaces';
import { getVisibleEditors } from '../../utils/editors-mosaic-arrangement';
import { AppState } from '../state';
import { TITLE_MAP } from './editors';

export interface EditorDropdownState {
  value: string;
}

export interface EditorDropdownProps {
  appState: AppState;
}

/**
 * A dropdown allowing users to toggle the various editors
 *
 * @class EditorDropdown
 * @extends {React.Component<EditorDropdownProps, EditorDropdownState>}
 */
@observer
export class EditorDropdown extends React.Component<EditorDropdownProps, EditorDropdownState> {
  constructor(props: EditorDropdownProps) {
    super(props);

    this.onItemClick = this.onItemClick.bind(this);
  }


  public render() {
    return (
      <Popover content={this.renderMenu()} position={Position.BOTTOM}>
        <Button icon='applications' text='Editors' />
      </Popover>
    );
  }

  public renderMenu() {
    return (
      <Menu>
        {...this.renderMenuItems()}
      </Menu>
    );
  }

  public renderMenuItems() {
    const { appState } = this.props;
    const editors = [ EditorId.main, EditorId.renderer, EditorId.html ];
    const result: Array<JSX.Element> = [];
    const visibleEditors = getVisibleEditors(appState.mosaicArrangement);

    for (const id of editors) {
      result.push(
        <MenuItem
          icon={visibleEditors.includes(id) ? 'eye-open' : 'eye-off'}
          key={id}
          text={TITLE_MAP[id]}
          id={id}
          onClick={this.onItemClick}
        />
      );
    }

    return result;
  }

  public onItemClick(event: React.MouseEvent) {
    const { id } = event.currentTarget;
    const { appState } = this.props;
    const visibleEditors = getVisibleEditors(appState.mosaicArrangement);

    if (visibleEditors.includes(id)) {
      appState.hideAndBackupEditor(id);
    } else {
      appState.showEditor(id);
    }
  }
}
