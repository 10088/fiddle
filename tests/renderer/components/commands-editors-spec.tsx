import { mount, shallow } from 'enzyme';
import * as React from 'react';

import { EditorId } from '../../../src/interfaces';
import { EditorDropdown } from '../../../src/renderer/components/commands-editors';
import { getVisibleEditors } from '../../../src/utils/editors-mosaic-arrangement';

jest.mock('../../../src/utils/editors-mosaic-arrangement');

describe('EditorDropdown component', () => {
  let store: any;

  beforeEach(() => {
    store = {
      hideAndBackupEditor: jest.fn(),
      showEditor: jest.fn()
    };

    (getVisibleEditors as jest.Mock).mockReturnValue([ EditorId.html, EditorId.renderer ]);
  });

  it('renders', () => {
    const wrapper = shallow(<EditorDropdown appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles a click for an item', () => {
    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    dropdown.onItemClick({ currentTarget: { id: EditorId.html } } as any);
    expect(store.hideAndBackupEditor).toHaveBeenCalledTimes(1);
    expect(store.showEditor).toHaveBeenCalledTimes(0);

    dropdown.onItemClick({ currentTarget: { id: EditorId.main } } as any);
    expect(store.hideAndBackupEditor).toHaveBeenCalledTimes(1);
    expect(store.showEditor).toHaveBeenCalledTimes(1);
  });
});
