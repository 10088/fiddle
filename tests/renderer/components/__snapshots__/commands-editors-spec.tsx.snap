// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorDropdown component renders 1`] = `
<Fragment>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <Blueprint3.Menu>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="main.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Main Process (main.js)"
        />
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="renderer.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Renderer Process (renderer.js)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="renderer.js"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="preload.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Preload (preload.js)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="preload.js"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="index.html"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="HTML (index.html)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="index.html"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="styles.css"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Stylesheet (styles.css)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="styles.css"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="plus"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Add New File"
          />
        </React.Fragment>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="grid-view"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Reset Layout"
          />
        </React.Fragment>
      </Blueprint3.Menu>
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="bottom"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Blueprint3.Button
      icon="applications"
      text="Editors"
    />
  </Blueprint3.Popover>
</Fragment>
`;

exports[`EditorDropdown component renders the extra button if the FIDDLE_DOCS_DEMOS is set 1`] = `
<Fragment>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <Blueprint3.Menu>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="main.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Main Process (main.js)"
        />
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="renderer.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Renderer Process (renderer.js)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="renderer.js"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="preload.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Preload (preload.js)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="preload.js"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="index.html"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="HTML (index.html)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="index.html"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="styles.css"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Stylesheet (styles.css)"
        >
          <Blueprint3.MenuItem
            disabled={false}
            icon="cross"
            id="styles.css"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Remove"
          />
        </Blueprint3.MenuItem>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="plus"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Add New File"
          />
        </React.Fragment>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="grid-view"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Reset Layout"
          />
        </React.Fragment>
      </Blueprint3.Menu>
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="bottom"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Blueprint3.Button
      icon="applications"
      text="Editors"
    />
  </Blueprint3.Popover>
</Fragment>
`;
